### PR TITLE
implement tagcolor features

### LINKED ISSUE #120: As a user, I want to set and change tag colors.
As a user, I want to set and change tag colors.

### PR DESCRIPTION
**Pull request by LuWenQ**
closes #120

### CODE REVIEW

File: src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+ */
+public class EditTagColorCommand extends Command {
+
+    public static final String COMMAND_WORD = "edittagcolor";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits the color of the Tag identified "
+            + "by the index number used in the displayed Tag list. "
+            + "Existing values will be overwritten by the input values.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + "[" + PREFIX_TAG_COLOR + "TAGCOLOR] \n"
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java]: This field shouldn't be optional.

File: src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            + "Example: " + COMMAND_WORD + " 1 " + PREFIX_TAG_COLOR + "BULE";
+    public static final String MESSAGE_EDIT_TAG_SUCCESS = "Edited Tag: %1$s";
+    public static final String MESSAGE_DUPLICATE_TAG = "This Tag already exists in the algobase.";
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java]: Why do we need this?

File: src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    public static final String COMMAND_WORD = "edittagcolor";
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Edits the color of the Tag identified "
+            + "by the index number used in the displayed Tag list. "
+            + "Existing values will be overwritten by the input values.\n"
+            + "Parameters: INDEX (must be a positive integer) "
+            + "[" + PREFIX_TAG_COLOR + "TAGCOLOR] \n"
+            + "Example: " + COMMAND_WORD + " 1 " + PREFIX_TAG_COLOR + "BULE";
+    public static final String MESSAGE_EDIT_TAG_SUCCESS = "Edited Tag: %1$s";
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java]: Follow the same format for the other command messages.

File: src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    private final Index index;
+    private final String color;
+
+    /**
+     * @param index of the Tag in the filtered Tag list to edit
+     * @param color details to edit the Tag with
+     */
+    public EditTagColorCommand(Index index, String color) {
+        requireNonNull(index);
+        requireNonNull(color);
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java]: Use `requireAllNonNull`.

File: src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        if (index.getZeroBased() >= lastShownList.size()) {
+            throw new CommandException(Messages.MESSAGE_INVALID_TAG_DISPLAYED_INDEX);
+        }
+
+        Tag tagToEdit = lastShownList.get(index.getZeroBased());
+        Tag editedTag = createEditedTag(tagToEdit, color);
+
+        if (!tagToEdit.isSameTag(editedTag) && model.hasTag(editedTag)) {
+            throw new CommandException(MESSAGE_DUPLICATE_TAG);
+        }
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java]: Hmm you are simply changing the colors, how can duplicates be possible? Duplicates are based on tag names only right?

File: src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    }
+
+    /**
+     * @param tagToEdit tag that need to edit
+     * @param color new tag name for tagToEdit
+     * @return Tag with updated name.
+     */
+    private static Tag createEditedTag(Tag tagToEdit, String color) {
+        assert tagToEdit != null;
+        assert color != null;
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/logic/commands/EditTagColorCommand.java]: Use `requireAllNonNull`.

**Review by jiayushe — CHANGES_REQUESTED** [review]
Hmm can't you just use the `edittag` command to change the colors? The user should be allowed to specify the tag color upon creating new tags as well.

File: src/test/java/seedu/algobase/logic/commands/HelpCommandTest.java [hunk 7]
```diff
@@ -19,6 +19,10 @@
 class HelpCommandTest {
     private static final CommandHistory EMPTY_COMMAND_HISTORY = new CommandHistory();
 
+    private static final String EXPECTED_COMMAND_LIST = "Available commands are: [add, addplan, addtag, "
```
Comment (jiayushe) [inline src/test/java/seedu/algobase/logic/commands/HelpCommandTest.java:22]: This is outdated and causing test failing.

File: src/main/java/seedu/algobase/model/tag/Tag.java [hunk 8]
```diff
@@ -45,13 +65,20 @@ public String getName() {
...
     /**
      * Returns true if a given string is a valid tag name.
      */
     public static boolean isValidTagName(String test) {
         return test.matches(VALIDATION_REGEX);
     }
 
+    public static boolean isValidTagColor(String color) {
+        return true;
+    }
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/tag/Tag.java]: What is the purpose of this method? Always return `true`?

File: src/main/java/seedu/algobase/model/tag/Tag.java [hunk 9]
```diff
@@ -28,13 +32,29 @@ public Tag(String tagName) {
         checkArgument(isValidTagName(tagName), MESSAGE_CONSTRAINTS);
         this.id = Id.generateId();
         this.tagName = tagName;
+        this.tagColor = DEFAULT_COLOR;
+    }
+    public Tag(String tagName, String tagColor) {
+        requireNonNull(tagName);
+        checkArgument(isValidTagName(tagName), MESSAGE_CONSTRAINTS);
```
Comment (jiayushe) [inline src/main/java/seedu/algobase/model/tag/Tag.java]: These 2 lines should be updated accordingly.

**Review by jiayushe — CHANGES_REQUESTED** [review]
Can you address my previous review as well? Thanks.

**Comment by LuWenQ** [comment]
Got bug in storage

File: docs/UserGuide.adoc [hunk 10]
```diff
@@ -286,6 +286,18 @@ Examples:
...
+
+Edits an existing tag's tag color +
+Format: `edittagcolor INDEX c/TAGCOLOR`
+
+Examples:
+
+* `edittagcolor 1 c/RED`
+
+[NOTE]
+Tag color can only be "RED" "ORANGE" "YELLOW" "GREEN" "BLUE" "PURPLE" "TEAL" "BLACK" and "DEFAULT".
```
Comment (le0tan) [inline docs/UserGuide.adoc]: **or** "DEFAULT"

**Review by le0tan — CHANGES_REQUESTED** [review]
Have you updated the UG for `addtag` and the `MESSAGE_USAGE` in `AddTagCommand`?
And can you check if your changes affect other parts of DG & UG?

**Review by le0tan — APPROVED** [review]

**Review by jiayushe — APPROVED** [review]
LGTM

**Pull request merged**